src/mlpack/core/tree/cellbound.hpp

```diff
@@ -0,0 +1,269 @@
+/**
+ * @file cellbound.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Definition of the CellBound class. The class describes a bound that consists
+ * of a number of hyperrectangles. These hyperrectangles do not overlap each
+ * other. The bound is limited by an outer hyperrectangle and two addresses,
+ * the lower address and the high address. Thus, the bound contains all points
+ * included between the lower and the high addresses.
+ *
+ * The notion of addresses is described in the following paper.
+ * @code
+ * @inproceedings{bayer1997,
+ *   author = {Bayer, Rudolf},
+ *   title = {The Universal B-Tree for Multidimensional Indexing: General
+ *       Concepts},
+ *   booktitle = {Proceedings of the International Conference on Worldwide
+ *       Computing and Its Applications},
+ *   series = {WWCA '97},
+ *   year = {1997},
+ *   isbn = {3-540-63343-X},
+ *   pages = {198--209},
+ *   numpages = {12},
+ *   publisher = {Springer-Verlag},
+ *   address = {London, UK, UK},
+ * }
+ * @endcode
+ */
+#ifndef MLPACK_CORE_TREE_CELLBOUND_HPP
+#define MLPACK_CORE_TREE_CELLBOUND_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/core/math/range.hpp>
+#include <mlpack/core/metrics/lmetric.hpp>
+#include "bound_traits.hpp"
+#include "address.hpp"
+
+namespace mlpack {
+namespace bound {
+
+template<typename MetricType = metric::LMetric<2, true>,
```

Comment:
Could you add some documentation here for the class itself?  You can reuse what you've written at the top of the file.  It might also be useful to point out that you are caching the minimum bounding rectangle of the points here.